File: Leetcode/String/424.py
class Solution:
    def canMakeValidSubstring(self, s: str, substring_length: int, k: int):
        freq_map = {}
        max_frequency = 0
        start = 0
        for end in range(len(s)):
            freq_map[s[end]] = freq_map.get(s[end], 0) + 1
            if end - start + 1 > substring_length:
                freq_map[s[start]] -= 1
                start += 1
        
            max_frequency = max(max_frequency, freq_map[s[end]])
            if substring_length - max_frequency <= k:
                return True
        
        return False

    def characterReplacement(self, s: str, k: int) -> int:
        lo = 1
        hi = len(s) + 1
        while lo + 1 < hi:
            mid = (lo + hi) // 2
            if self.canMakeValidSubstring(s, mid, k):
                lo = mid
            else:
                hi = mid
        return lo

File: Leetcode/String/424.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function (s, k) {
  let map = {};
  let start = 0;
  let maxFreq = 0;
  let longest = 0;

  for (let end = 0; end < s.length; end++) {
    map[s[end]] = (map[s[end]] || 0) + 1;
    maxFreq = Math.max(maxFreq, map[s[end]]);

    if (end - start + 1 - maxFreq > k) {
      map[s[start]]--;
      start++;
    } else {
      longest = Math.max(longest, end - start + 1);
    }
  }

  return longest;
};
